core/src/test/java/org/apache/calcite/test/enumerable/EnumerableHashJoinTest.java

```diff
@@ -74,6 +74,25 @@ class EnumerableHashJoinTest {
             "empid=110; name=Theodore; dept=Sales");
   }
 
+  @Test void innerJoinWithCompositeKeyAndNullValues() {
+    tester(false, new HrSchema())
+        .query(
+            "select e1.empid from emps e1 join emps e2 "
+                + "on e1.deptno=e2.deptno and e1.commission=e2.commission")
+        .withHook(Hook.PLANNER, (Consumer<RelOptPlanner>) planner ->
+            planner.removeRule(EnumerableRules.ENUMERABLE_MERGE_JOIN_RULE))
+        .explainContains("EnumerableCalc(expr#0..4=[{inputs}], empid=[$t0])\n"
+            + "  EnumerableHashJoin(condition=[AND(=($1, $3), =($2, $4))], joinType=[inner])\n"
```

Comment:
Do we need test cases for outer join with composite keys?

Reply:
Good point. More tests added.